workers/deployment_manager/DeploymentManager/Worker/DeploymentManager.cs

```diff
@@ -0,0 +1,187 @@
+            };
+
+            var connector = new Connector(WorkerType);
+            metaConnection = connector.TryToConnect(connectionParameters, options.ReceptionistHost, options.ReceptionistPort, options.WorkerId);
+
+            devAuthToken = WorkerAuthenticator.RetrieveDevelopmentAuthenticationToken(options.ProjectName);
+        }
+
+        private string GetFileContent(string fileName)
+        {
+            var result = string.Empty;
+            var assembly = Assembly.GetExecutingAssembly();
+            using (Stream stream = assembly.GetManifestResourceStream($"DeploymentManager.{fileName}"))
+            using (StreamReader reader = new StreamReader(stream))
+            {
+                return reader.ReadToEnd();
+            }
+        }
+
+        public void ObserveDeployments()
+        {
+            // spin up the other deployments
+            for (var i = 0; i < options.NumberOfDeployments; i++)
+            {
+                // TODO check if deployment is already running
+                var deploymentName = $"{options.DeploymentPrefix}_{i}";
+                new Thread(new ParameterizedThreadStart(StartDeployment)).Start(deploymentName);
+            }
+
+            while (metaConnection.GetConnectionStatusCode() == ConnectionStatusCode.Success)
+            {
+                using (var dispatcher = new Dispatcher())
+                using (var opList = metaConnection.GetOpList(0))
+                {
+                    dispatcher.Process(opList);
+                }
+            }
+
+            Log.Print(LogLevel.Info, "Disconnected from SpatialOS");
+        }
+
+        private void StartDeployment(Object obj)
+        {
+            try
+            {
+                string deploymentName = (string)obj;
+
+                var snapshotPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "default.snapshot");
+                var launchConfig = new LaunchConfig
+                {
+                    ConfigJson = GetFileContent("launch_config.json"),
+                };
+
+                Log.Print(LogLevel.Info, $"Uploading {snapshotPath} to project {options.ProjectName}", metaConnection);
+                var snapshotId = DeploymentModifier.UploadSnapshot(snapshotPath, options.ProjectName, deploymentName);
+
+                if (snapshotId.Length == 0)
+                {
+                    Log.Print(LogLevel.Error, "Something went wrong during snapshot upload.", metaConnection);
+                    return;
+                }
+
+                var template = new DeploymentTemplate
+                {
+                    AssemblyId = options.AssemblyName,
+                    DeploymentName = deploymentName,
+                    LaunchConfig = launchConfig,
+                    ProjectName = options.ProjectName,
+                    SnapshotId = snapshotId,
+                };
+
+                while (metaConnection.GetConnectionStatusCode() == ConnectionStatusCode.Success)
```

Comment:
Is this allowed? The connection isn't generally thread safe.

Reply:
It should be "fine", but the connection APIs are pretty unclear...

Reply:
My understanding is its _fine_ to access the connection from different threads, but no concurrent operations are allowed.

Reply:
had an offline discussion: will look into how to refactor this so that only starting the deployments is in a separate thread and polling will be done on the main thread. that might be a better solution